src/validate.ts

```diff
@@ -0,0 +1,70 @@
+import type * as typescript from "typescript";
+import { FunctionlessChecker } from "./checker";
+
+/**
+ * Validates a TypeScript SourceFile containing Functionless primitives does not
+ * unsupported syntax.
+ *
+ * @param ts a reference to the TypeScript Server.
+ * @param checker the Program's {@link FunctionlessChecker}.
+ * @param sf the TypeScript SourceFile.
+ * @returns diagnostic errors for the file.
+ */
+export function validate(
+  ts: typeof typescript,
+  checker: FunctionlessChecker,
+  node: typescript.Node,
+  logger?: {
+    info(message: string): void;
+  }
+): typescript.Diagnostic[] {
+  logger?.info("Beginning validation of Functionless semantics");
+
+  return (function visit(node: typescript.Node): typescript.Diagnostic[] {
+    return [
+      ...(checker.isStepFunction(node) ? validateStepFunctionNode(node) : []),
+      ...(visitEachChild(node, visit) ?? []),
+    ];
+  })(node);
+
+  function validateStepFunctionNode(
+    node: typescript.Node
+  ): typescript.Diagnostic[] {
+    const children = visitEachChild(node, validateStepFunctionNode);
+    if (ts.isBinaryExpression(node)) {
+      if (
+        [ts.SyntaxKind.PlusToken, ts.SyntaxKind.MinusToken].includes(
+          node.operatorToken.kind
+        )
+      ) {
+        return [
+          <typescript.Diagnostic>{
+            source: "Functionless",
+            key: "Functionless",
+            category: ts.DiagnosticCategory.Error,
+            code: 100,
+            file: node.getSourceFile(),
+            start: node.pos,
+            length: node.end - node.pos,
+            messageText: "arithmetic is not supported in a Step Function",
+          },
+        ].concat(children);
+      }
+    }
+    return children;
+  }
+
+  // ts.forEachChild terminates whenever a truth value is returned
+  // ts.visitEachChild requires a ts.TransformationContext, so we can't use that
+  // this wrapper uses a mutable array to collect the results
+  function visitEachChild<T>(
```

Comment:
`validateEachChild`?

The behavior is different from `visitEachChild` which returns nodes back into the location of the node being processed.

Reply:
Or map/reduce eachChild?

Reply:
`valdiateEachChild` seems fine enough. I also struggled with the name.